code/game/machinery/computer/launchpad_control.dm

```diff
@@ -12,6 +12,102 @@
 /obj/machinery/computer/launchpad/Initialize()
 	launchpads = list()
 	. = ..()
+	AddComponent(/datum/component/usb_port, list(
+		/obj/item/circuit_component/bluespace_launchpad,
+	))
+
+/obj/item/circuit_component/bluespace_launchpad
+	display_name = "Bluespace Launchpad Console"
+	display_desc = "Teleports anything to and from any location on the station. Doesn't use actual GPS coordinates, but rather offsets from the launchpad itself. And it can only go as far as the launchpad can go, which depends on its components."
+
+	var/datum/port/input/launchpad_id
+	var/datum/port/input/x_pos
+	var/datum/port/input/y_pos
+	var/datum/port/input/send_trigger
+	var/datum/port/input/retrieve_trigger
+
+	var/datum/port/output/sent
+	var/datum/port/output/retrieved
+	var/datum/port/output/on_fail
+	var/datum/port/output/why_fail
+
+	var/obj/machinery/computer/launchpad/attached_console
+
+/obj/item/circuit_component/bluespace_launchpad/Initialize()
+	. = ..()
+	launchpad_id = add_input_port("Launchpad ID", PORT_TYPE_NUMBER, FALSE)
+	x_pos = add_input_port("X offset", PORT_TYPE_NUMBER, FALSE)
+	y_pos = add_input_port("Y offset", PORT_TYPE_NUMBER, FALSE)
+	send_trigger = add_input_port("Send", PORT_TYPE_SIGNAL)
+	retrieve_trigger = add_input_port("Retrieve", PORT_TYPE_SIGNAL)
+
+	sent = add_output_port("Sent", PORT_TYPE_SIGNAL)
+	retrieved = add_output_port("Retrieved", PORT_TYPE_SIGNAL)
+	why_fail = add_output_port("Fail reason", PORT_TYPE_STRING)
+	on_fail = add_output_port("Failed", PORT_TYPE_SIGNAL)
+
+/obj/item/circuit_component/bluespace_launchpad/Destroy()
+	launchpad_id = null
+	x_pos = null
+	y_pos = null
+	send_trigger = null
+	retrieve_trigger = null
+	sent = null
+	retrieved = null
+	on_fail = null
+	why_fail = null
+	return ..()
+
+/obj/item/circuit_component/bluespace_launchpad/register_usb_parent(atom/movable/parent)
+	. = ..()
+	if(istype(parent, /obj/machinery/computer/launchpad))
+		attached_console = parent
+
+/obj/item/circuit_component/bluespace_launchpad/unregister_usb_parent(atom/movable/parent)
+	attached_console = null
+	return ..()
+
+/obj/item/circuit_component/bluespace_launchpad/input_received(datum/port/input/port)
+	. = ..()
+
+	if(.)
+		return
+
+	if(!attached_console || length(attached_console.launchpads) == 0)
+		on_fail.set_output(COMPONENT_SIGNAL)
+		why_fail.set_output("No launchpads connected!")
```

Comment:
you probably don't need this but it doesn't hurt.
I'll soon implement an easier way to flag up errors rather than having to do this.
Also you need to send the why_fail first, on_fail comes last.

Reply:
Well now that there are WAY more possible error messages I think it's kinda necesssary lol